dandi/validate.py

```diff
@@ -44,12 +90,101 @@ def validate_bids(
+        version=bidsschematools.__version__,
+    )
+
+    for path in validation_result["path_tracking"]:
+        # Hard-coding exclusion here pending feature + release in:
+        # https://github.com/bids-standard/bids-specification/issues/1272
+        if path.endswith((".ERRORS", ".ERRORS.json")):
+            continue
+        dataset_path = _get_set_path(path, "dataset_description.json")
+        dandiset_path = _get_set_path(path, "dandiset.yaml")
+        our_validation_result.append(
+            ValidationResult(
+                origin=origin,
+                severity=Severity.ERROR,
+                id="BIDS.NON_BIDS_PATH_PLACEHOLDER",
+                scope=Scope.FILE,
+                path=Path(path),
+                message="File does not match any pattern known to BIDS.",
+                dataset_path=dataset_path,
+                dandiset_path=dandiset_path,
+            )
+        )
+
+    for pattern in validation_result["schema_tracking"]:
+        # Future proofing for standard-compliant name.
+        if pattern.get("mandatory") or pattern.get("required"):
+            # We don't have a path for this so we'll need some external logic to make sure
+            # that the dataset path is populated.
+            # dataset_path = _get_dataset_path(path, paths)
+            our_validation_result.append(
+                ValidationResult(
+                    origin=origin,
+                    severity=Severity.ERROR,
+                    id="BIDS.MANDATORY_FILE_MISSING_PLACEHOLDER",
+                    scope=Scope.DATASET,
+                    path_regex=pattern["regex"],
+                    message="BIDS-required file is not present.",
+                )
+            )
+    for meta in validation_result["match_listing"]:
+        file_path = meta.pop("path")
+        meta = {BIDS_TO_DANDI[k]: v for k, v in meta.items() if k in BIDS_TO_DANDI}
+        dataset_path = _get_set_path(file_path, "dataset_description.json")
+        dandiset_path = _get_set_path(file_path, "dandiset.yaml")
+        # Top level files do not have any other metadata other than path,
+        # which we pop and put in the object...
+        if not meta:
+            meta = {
+                "subject": None,
+            }
+        our_validation_result.append(
+            ValidationResult(
+                origin=origin,
+                id="BIDS.MATCH",
+                scope=Scope.FILE,
+                path=Path(file_path),
+                metadata=meta,
+                dataset_path=dataset_path,
+                dandiset_path=dandiset_path,
+            )
+        )
+
+    return our_validation_result
+
+
+def _get_set_path(in_path, marker):
+    """Get the path to the root of a file set (e.g. BIDS dataset or DANDIset).
+
+    Parameters
+    ----------
+    in_path : str, optional
+        File for which to determine set path.
+    marker : str, optional
+        Filename marker which identifies the set root.
+    """
+    import os
```

Comment:
in general please define such generic import to top of the file

Reply:
3acc2c93cda314f6326985999b2db6f8786de47e

Reply:
ultimately not needed, actually.

b76b8bbbd94dfa3ebfc969ed08482369bf55d28b